File: app/static/intelligence.js
let attentionCards = [];
let performanceData;
let performanceGroup = 'sources';
const categoryNames = {reply:'Messages', accepted:'Booking fees', new:'New enquiries', interest:'Recent interest', waiting:'Waiting quotes'};

async function renderToday() {
  const data = await api('/api/admin/intelligence/today');
  attentionCards = data.cards;
  workspace.className = 'workspace';
  workspace.innerHTML = `<div class="page-heading"><div><small class="eyebrow">YOUR MORNING CHECK-IN</small><h1>Who needs me today?</h1><p>${data.cards.length} couples to review · updated ${niceTime(data.generated_at)}</p></div><button class="secondary" id="refresh-today">Refresh</button></div>
    ${data.missing_evidence ? `<div class="notice">${data.missing_evidence} records are waiting for the new Booking activity connector. Recommendations appear when their evidence arrives.</div>` : ''}
    <div class="notice">Mailbox evidence covers the latest 200 inbox headers, matched by the main email address. Archived mail, partner addresses, phone calls and WhatsApp may be missing. Review the conversation before contacting a couple.</div>
    <div class="attention-filters"><button class="secondary active" data-category="all">All · ${data.cards.length}</button>${Object.entries(categoryNames).map(([k,v])=>`<button class="secondary" data-category="${k}">${v} · ${data.cards.filter(c=>c.kind===k).length}</button>`).join('')}</div>
    ${data.snoozed ? `<p>${data.snoozed} cards snoozed <button class="secondary" id="restore-snoozes">Show again</button></p>` : ''}
    <div class="attention-grid" id="attention-cards"></div>`;
  document.querySelector('#refresh-today').onclick = () => renderToday().catch(e=>toast(e.message));
  document.querySelector('#restore-snoozes')?.addEventListener('click',async()=>{try{await api('/api/admin/intelligence/reset-snoozes',{method:'POST'});await renderToday()}catch(e){toast(e.message)}});
  document.querySelectorAll('[data-category]').forEach(b=>b.onclick=()=>{
    document.querySelectorAll('[data-category]').forEach(x=>x.classList.toggle('active',x===b));drawAttention(b.dataset.category);
  });
  drawAttention('all');
}

function drawAttention(category) {
  const rows=attentionCards.filter(c=>category==='all'||c.kind===category);
  document.querySelector('#attention-cards').innerHTML=rows.length ? rows.map(c=>`<article class="attention-card ${esc(c.kind)}" data-card="${esc(c.lead_id)}">
    <header><span class="status">${esc(categoryNames[c.kind])}</span><strong>${money(c.value)}</strong></header>
    <h2>${esc(c.couple_name)}</h2><p class="muted">${niceDate(c.event_date)} · ${esc(c.venue)}</p>
    <h3>${esc(c.title)}</h3><p>${esc(c.reason)}</p>
    ${c.stale?'<div class="notice">Booking evidence is over 30 minutes old. Check Booking for the latest position.</div>':''}
    <details><summary>Why this is here</summary><dl class="evidence">
      <dt>Quote sent</dt><dd>${niceTime(c.quote_sent_at)}</dd><dt>Last outgoing contact</dt><dd>${niceTime(c.last_contact_at)}</dd>
      <dt>Incoming email seen</dt><dd>${niceTime(c.last_incoming_at)}</dd><dt>Quote-link access</dt><dd>${niceTime(c.quote_link_at)}</dd>
      <dt>Mailbox check</dt><dd>${esc({recent_inbox:'Recent inbox headers checked',unavailable:'Mailbox unavailable',not_configured:'Mailbox not configured'}[c.mail_status])} · ${niceTime(c.mail_checked_at)}</dd>
      <dt>Booking synced</dt><dd>${niceTime(c.synced_at)}</dd></dl></details>
    ${c.draft?`<details class="reply-draft"><summary>Review a suggested reply</summary><p>Prepared from the recorded stage and wedding details. Edit it, copy it, then review and send in Booking.</p><label>Message<textarea rows="8">${esc(c.draft)}</textarea></label><button class="secondary" data-copy="${esc(c.lead_id)}">Copy message</button></details>`:''}
    <footer><a class="primary button-link" href="${esc(c.url)}" target="_blank" rel="noopener">Open in Booking</a><button class="secondary" data-snooze="${esc(c.lead_id)}">Tomorrow</button><button class="secondary" data-lead="${esc(c.lead_id)}">Details</button></footer>
  </article>`).join('') : '<div class="panel empty">Nothing to review in this group. Snoozed cards return automatically when their time is up.</div>';
  bindLeadRows();
  document.querySelectorAll('[data-copy]').forEach(b=>b.onclick=async()=>{
    const textarea=b.closest('article').querySelector('textarea');
    try{await navigator.clipboard.writeText(textarea.value);toast('Copied—review and send in Booking')}catch{textarea.select();toast('Select and copy the message from the text box')}
  });
  document.querySelectorAll('[data-snooze]').forEach(b=>b.onclick=async()=>{
    try{await api(`/api/admin/intelligence/${b.dataset.snooze}/snooze`,{method:'POST',body:JSON.stringify({hours:24})});await renderToday()}catch(e){toast(e.message)}
  });
}

async function renderPerformance() {
  workspace.className='workspace';
  workspace.innerHTML=`<div class="page-heading"><div><small class="eyebrow">WHERE YOUR BOOKINGS COME FROM</small><h1>What’s working?</h1><p>Follow enquiries through to confirmed bookings.</p></div></div>
    <form class="toolbar" id="performance-filter"><label>Enquiries received from <input type="date" name="start"></label><label>To <input type="date" name="end"></label><button class="primary">Apply dates</button><button class="secondary" type="button" id="all-dates">All synced dates</button></form>
    <p class="muted">Dates refer to the original enquiry, not the wedding or import date. This covers only records included in your Booking sync range. Test and archived records are excluded. Booked value is agreed work, not cash collected.</p><div id="performance-content"></div>`;
  document.querySelector('#performance-filter').onsubmit=e=>{e.preventDefault();loadPerformance().catch(e=>toast(e.message))};
  document.querySelector('#all-dates').onclick=()=>{document.querySelector('#performance-filter').reset();loadPerformance().catch(e=>toast(e.message))};
  await loadPerformance();
}

async function loadPerformance() {
  const fields=new FormData(document.querySelector('#performance-filter'));
  const query=new URLSearchParams([...fields].filter(([,v])=>v));
  performanceData=await api(`/api/admin/intelligence/performance?${query}`);
  const d=performanceData;
  document.querySelector('#performance-content').innerHTML=`${d.missing_evidence?`<div class="notice">${d.missing_evidence} records lack source-date evidence and are excluded until the updated sync.</div>`:''}
    <section class="metrics"><article class="metric"><label>Enquiries in range</label><strong>${d.enquiries}</strong></article><article class="metric"><label>Confirmed bookings</label><strong>${d.bookings}</strong><small>${d.conversion}% of enquiries</small></article><article class="metric"><label>Booked value</label><strong>${money(d.booked_value)}</strong></article><article class="metric"><label>Average booking</label><strong>${money(d.average_booking)}</strong></article></section>
    <div class="attention-filters">${Object.entries({sources:'Enquiry sources',venues:'Venues',packages:'Packages',addons:'Add-ons'}).map(([k,v])=>`<button class="secondary" data-group="${k}">${v}</button>`).join('')}</div><div id="performance-table"></div>`;
  document.querySelectorAll('[data-group]').forEach(b=>b.onclick=()=>{performanceGroup=b.dataset.group;drawPerformance()});drawPerformance();
}

function drawPerformance() {
  document.querySelectorAll('[data-group]').forEach(b=>b.classList.toggle('active',b.dataset.group===performanceGroup));
  const rows=performanceData.groups[performanceGroup], addon=performanceGroup==='addons';
  const best=rows.find(r=>r.bookings>0 && (addon || !r.small_sample));
  document.querySelector('#performance-table').innerHTML=`<p>${addon?'Selected add-ons on confirmed bookings. Counts are distinct bookings; amounts are add-on line totals. Discounts are excluded.':'Conversion is confirmed bookings divided by enquiries in the selected range. Groups below 10 enquiries are marked as a small sample.'}</p>
    ${performanceGroup==='packages'?'<p class="muted">Package groups use the selected quote package, falling back to recorded interest. They show associations, not proof that a package caused a booking.</p>':''}
    ${best?`<div class="notice"><strong>${esc(best.name)}</strong> has the highest booked value ${addon?'among selected add-ons':'among groups with at least 10 enquiries'}: ${money(best.booked_value)} from ${best.bookings} bookings.</div>`:''}
    ${rows.length?`<div class="table-wrap"><table class="insight-table"><thead><tr><th>Name</th>${addon?'':'<th>Enquiries</th>'}<th>Bookings</th>${addon?'':'<th>Conversion</th>'}<th>${addon?'Add-on value':'Booked value'}</th></tr></thead><tbody>${rows.map(r=>`<tr><td><strong>${esc(r.name)}</strong>${r.small_sample?'<small>Small sample</small>':''}</td>${addon?'':`<td>${r.enquiries}</td>`}<td>${r.bookings}</td>${addon?'':`<td><span>${r.conversion}%</span><progress max="100" value="${r.conversion}" aria-label="Conversion ${r.conversion} percent"></progress></td>`}<td>${money(r.booked_value)}</td></tr>`).join('')}</tbody></table></div>`:'<div class="panel empty">No results in this group for the selected dates.</div>'}`;
}

File: app/intelligence.py
"""Evidence-based priorities and cohort reporting; never sends messages."""
from collections import defaultdict
from datetime import date, datetime, timedelta, timezone
from urllib.parse import quote
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.orm import Session
from .config import get_settings
from .db import get_db
from .models import BookingInsight, Lead
from .security import current_admin, csrf_admin

router = APIRouter(prefix='/api/admin/intelligence')


def utc(value):
    if not value:
        return None
    if isinstance(value, str):
        value = datetime.fromisoformat(value.replace('Z', '+00:00'))
    return value.replace(tzinfo=timezone.utc) if value.tzinfo is None else value.astimezone(timezone.utc)


def records(db):
    return db.execute(select(Lead, BookingInsight).outerjoin(BookingInsight, BookingInsight.lead_id == Lead.id)
                      .where(Lead.external_booking_id.is_not(None))).all()


def active_record(f):
    return not (f.get('is_test') or f.get('archived'))


def won(f):
    return f['booking_status'] != 'cancelled' and (f['booking_status'] in {'confirmed', 'in_progress', 'completed'} or f.get('deposit_paid'))


def priority(lead, insight, now):
    if not insight or not active_record(insight.facts) or insight.facts.get('suppressed'):
        return None
    f = insight.facts
    if won(f) or f['booking_status'] == 'cancelled' or lead.event_date < now.date():
        return None
    if utc(insight.snoozed_until) and utc(insight.snoozed_until) > now:
        return None
    sent, contacted, incoming = (utc(f.get(k)) for k in ('quote_sent_at', 'last_contact_at', 'last_incoming_at'))
    clicked, created = utc(f.get('quote_link_at')), utc(f['source_created_at'])
    section, draft = 'journey', ''
    greeting = f'Hi {lead.primary_first_name} & {lead.partner_first_name},'
    if incoming and incoming >= created and (not contacted or incoming > contacted):
        kind, title, section = 'reply', 'A message needs a look', 'activity'
        reason = 'An incoming email from this address is newer than the last recorded outgoing message. Check the conversation before replying.'
    elif f.get('quote_accepted'):
        kind, title, section = 'accepted', 'Quote accepted · check the booking fee', 'payments'
        reason = 'An accepted quote has no first payment recorded. Check your bank before contacting the couple.'
        draft = f'{greeting}\n\nThank you for accepting your quote. If you have any questions about the booking fee or the next steps, just let me know and I’ll be happy to help.\n\nMark'
    elif not sent and f['booking_status'] == 'enquiry':
        kind, title = 'new', 'Prepare their quote'
        reason = f'Enquiry received {created.strftime("%d %b %Y")}; no successful quote email is recorded in Booking.'
    elif clicked and now - clicked <= timedelta(days=1) and (not contacted or clicked > contacted):
        kind, title = 'interest', 'Recent quote-link activity'
        reason = 'A quote-link access was recorded in the last 24 hours. This is an interest signal, not proof a person read it; automated scanners can open links.'
        draft = f'{greeting}\n\nI hope the plans for your wedding at {lead.venue} are coming together nicely. Is there anything you’d like to ask about the coverage or how I work on the day?\n\nMark'
    elif sent and now - max(sent, contacted or sent) >= timedelta(days=3):
        kind, title = 'waiting', 'Quote waiting · review the conversation'
        reason = 'At least three days since the last recorded outgoing contact. Check for replies or contact outside Booking before following up.'
        draft = f'{greeting}\n\nI just wanted to check whether you had any questions about the wedding information I sent for {lead.venue}. There’s absolutely no pressure—if there’s anything I can help with, just let me know.\n\nMark'
    else:
        return None
    return {'lead_id': lead.id, 'couple_name': lead.couple_name, 'event_date': lead.event_date.isoformat(),
            'venue': lead.venue, 'kind': kind, 'title': title, 'reason': reason,
            'value': float(lead.estimated_value or 0), 'draft': draft,
            'url': get_settings().booking_action_url.rstrip('/') + '/bookings/' + quote(lead.external_booking_id, safe='') + '/' + section,
            'mail_status': f.get('mail_status', 'not_configured'), 'mail_checked_at': f.get('mail_checked_at'),
            'last_incoming_at': f.get('last_incoming_at'), 'last_contact_at': f.get('last_contact_at'),
            'quote_sent_at': f.get('quote_sent_at'), 'quote_link_at': f.get('quote_link_at'),
            'synced_at': utc(insight.received_at).isoformat(),
            'stale': now - utc(insight.received_at) > timedelta(minutes=30)}


@router.get('/today')
def today(_=Depends(current_admin), db: Session = Depends(get_db)):
    now = datetime.now(timezone.utc)
    rows = records(db)
    cards = [card for lead, insight in rows if (card := priority(lead, insight, now))]
    order = {'reply': 0, 'accepted': 1, 'new': 2, 'interest': 3, 'waiting': 4}
    cards.sort(key=lambda c: (order[c['kind']], c['event_date'], c['couple_name']))
    return {'cards': cards, 'missing_evidence': sum(1 for _, i in rows if not i),
            'snoozed': sum(1 for _, i in rows if i and utc(i.snoozed_until) and utc(i.snoozed_until) > now),
            'generated_at': now.isoformat()}


class SnoozeIn(BaseModel):
    hours: int = Field(default=24, ge=0, le=168)


@router.post('/{lead_id}/snooze')
def snooze(lead_id: str, payload: SnoozeIn, _=Depends(csrf_admin), db: Session = Depends(get_db)):
    row = db.get(BookingInsight, lead_id)
    if not row:
        raise HTTPException(404, 'Booking evidence not received yet')
    row.snoozed_until = datetime.now(timezone.utc) + timedelta(hours=payload.hours) if payload.hours else None
    db.commit()
    return {'ok': True}


@router.post('/reset-snoozes')
def reset_snoozes(_=Depends(csrf_admin), db: Session = Depends(get_db)):
    for row in db.scalars(select(BookingInsight).where(BookingInsight.snoozed_until.is_not(None))):
        row.snoozed_until = None
    db.commit()
    return {'ok': True}


def performance_rows(rows, start=None, end=None):
    groups = {k: {} for k in ('sources', 'venues', 'packages', 'addons')}
    included, missing = [], 0
    for lead, insight in rows:
        if not insight:
            missing += 1
            continue
        f = insight.facts
        if not active_record(f):
            continue
        created = utc(f['source_created_at']).date()
        if (start and created < start) or (end and created > end):
            continue
        included.append((lead, f))
        for group, label in [('sources', lead.referral_source or 'Not recorded'), ('venues', lead.venue or 'Not recorded'),
                             ('packages', next((x.get('name') for x in lead.quote_items or [] if x.get('type') == 'package'), None) or lead.package_interest or 'Not selected')]:
            label = label.strip() or 'Not recorded'
            row = groups[group].setdefault(label.casefold(), {'name': label, 'enquiries': 0, 'bookings': 0, 'booked_value': 0.0})
            row['enquiries'] += 1
            if won(f):
                row['bookings'] += 1
                row['booked_value'] += float(lead.estimated_value or 0)
        if won(f):
            selected = defaultdict(float)
            for item in lead.quote_items or []:
                if item.get('type') == 'addon':
                    selected[str(item.get('name') or 'Unnamed add-on').strip()] += float(item.get('total') or 0)
            for name, value in selected.items():
                row = groups['addons'].setdefault(name.casefold(), {'name': name, 'bookings': 0, 'booked_value': 0.0})
                row['bookings'] += 1
                row['booked_value'] += value
    for group in ('sources', 'venues', 'packages'):
        for row in groups[group].values():
            row['conversion'] = round(row['bookings'] / row['enquiries'] * 100, 1)
            row['small_sample'] = row['enquiries'] < 10
            row['booked_value'] = round(row['booked_value'], 2)
    wins = [l for l, f in included if won(f)]
    total = sum(float(l.estimated_value or 0) for l in wins)
    return {'groups': {k: sorted(v.values(), key=lambda r: (-r['booked_value'], r['name'])) for k, v in groups.items()},
            'enquiries': len(included), 'bookings': len(wins), 'booked_value': round(total, 2),
            'conversion': round(len(wins) / len(included) * 100, 1) if included else 0,
            'average_booking': round(total / len(wins), 2) if wins else 0, 'missing_evidence': missing}


@router.get('/performance')
def performance(start: date | None = None, end: date | None = None, _=Depends(current_admin), db: Session = Depends(get_db)):
    if start and end and start > end:
        raise HTTPException(422, 'The start date must be before the end date')
    return performance_rows(records(db), start, end)
